primitives/src/crh/poseidon/mod.rs

```diff
@@ -1,427 +1,213 @@
-                    vec_state[i][j] = w_bar * &w[idx as usize];
-                    w_bar = w_bar * &vec_1;
-                    idx -= 1;
-                }
-            }
+            let padding = if personalization.len() % P::R != 0 {
+                P::R - ( personalization.len() % P::R )
+            } else {
+                0
+            };
 
+            for _ in 0..padding {
+                instance.update(F::zero());
+            }
+            assert_eq!(instance.pending.len(), 0);
+            instance.updates_ctr = 0;
         }
+        instance
     }
 
-    fn poseidon_partial_round(vec_state: &mut Vec<Vec<P::Fr>>, round_cst_idx: &mut usize) {
-
-        // For each of the state vector element position
-        for j in 0..P::T {
-
-            // get the constant associated to state vector element position
-            let rc = P::ROUND_CST[*round_cst_idx];
-
-            // go over each state vector
-            for k in 0..vec_state.len() {
-                vec_state[k][j] += &rc;
+    #[inline]
+    fn apply_permutation_if_needed(&mut self) {
+        if self.pending.len() == P::R {
+            for (input, state) in self.pending.iter().zip(self.state.iter_mut()) {
+                *state += input;
             }
-            *round_cst_idx += 1;
+            Self::poseidon_perm(&mut self.state);
+            self.pending.clear();
         }
+    }
 
-        // Apply the S-BOX to the first elements of each of the state vector
-        let mut w: Vec<P::Fr> = Vec::new();
-        let mut accum_prod = P::Fr::one();
-
-        w.push(accum_prod);
-        // Calculate the intermediate partial products
-        for i in 0..vec_state.len() {
-            accum_prod = accum_prod * &vec_state[i][0];
-            w.push(accum_prod);
-        }
-
-        // if the accum_prod is zero, it means that one of the S-Boxes is zero
-        // in that case compute the inverses individually
-        if accum_prod == P::Fr::zero() {
-            for i in 0..(vec_state.len() - 1) {
-                if vec_state[i][0] != P::Fr::zero() {
-                    vec_state[i][0] = vec_state[i][0].inverse().unwrap();
-                }
-            }
-        } else {
-
-            // Calculate the inversion of the products
-            // Use Montgomery simultaneous inversion
-            let mut w_bar = accum_prod.inverse().unwrap();
-
-            // Extract the individual inversions
-            let mut idx: i64 = w.len() as i64 - 2;
-            for i in (0..vec_state.len()).rev() {
-                let vec_1 = vec_state[i][0].clone();
-                vec_state[i][0] = w_bar * &w[idx as usize];
-                w_bar = w_bar * &vec_1;
-                idx -= 1;
-            }
+    #[inline]
+    fn get_hash(mut state: Vec<F>, inputs: Vec<F>) -> F {
+        for (input, s) in inputs.iter().zip(state.iter_mut()) {
+            *s += input;
         }
+        Self::poseidon_perm(&mut state);
+        state[0]
```

Comment:
In general, one is allowed to return capacity many elements (from the rate-part of the state).